### PR TITLE
wrap: working to improve security

### PR DESCRIPTION
**Pull request by scivision**
### use WrapDB domain whitelist, don't fallback to non-SSL when SSL available

In my opinion, we should not fall back to http:// from the SSL HSTS WrapDB URL, **for systems that have Python SSL** as that is controverting the point of HSTS + SSL.

For systems that do not have Python SSL, they continue to work with a
colored mlog.warning instead of only a stderr console print.

Also, attempt to stop masquerade URLS containing wrapdb.mesonbuild.com.evil.stuff.com

### CODE REVIEW

File: mesonbuild/wrap/wrap.py [hunk 1]
```diff
@@ -65,24 +69,26 @@ def quiet_git(cmd: typing.List[str], workingdir: str) -> typing.Tuple[bool, str]
 
 def open_wrapdburl(urlstring: str) -> 'http.client.HTTPResponse':
     global ssl_warning_printed
+
     if has_ssl:
+        if not urlstring.startswith(whitelist_domain):
+            raise WrapException('{} is not a whitelisted URL'.format(urlstring))
```
Comment (jpakkane) [inline mesonbuild/wrap/wrap.py]: There is work afoot to create wrapdb mirrors, that most likely will have names such as `mirror1.wrapdb.mesonbuild.com`. For future proofness, this code should parse the server name and validate that it is a subdomain of wrapdb.

File: mesonbuild/wrap/wrap.py [hunk 2]
```diff
@@ -65,24 +69,26 @@ def quiet_git(cmd: typing.List[str], workingdir: str) -> typing.Tuple[bool, str]
...
-            if not ssl_warning_printed:
-                print('SSL connection failed. Falling back to unencrypted connections.', file=sys.stderr)
-                ssl_warning_printed = True
+        except urllib.error.URLError as excp:
+            raise WrapException('WrapDB connection failed to {} with error {}'.format(urlstring, excp))
+
+    # following code is only for those without Python SSL
+    nossl_urlstring = urlstring.replace('https://', 'http://')
+    if not nossl_urlstring.startswith(whitelist_domain_nossl):
+        raise WrapException('{} is not a whitelisted URL'.format(nossl_urlstring))
```
Comment (jpakkane) [inline mesonbuild/wrap/wrap.py]: Same here.

File: mesonbuild/wrap/wrap.py [hunk 3]
```diff
@@ -308,8 +314,10 @@ def get_data(self, url: str) -> typing.Tuple[str, str]:
         blocksize = 10 * 1024
         h = hashlib.sha256()
         tmpfile = tempfile.NamedTemporaryFile(mode='wb', dir=self.cachedir, delete=False)
-        if url.startswith('https://wrapdb.mesonbuild.com'):
+        if url.startswith((whitelist_domain, whitelist_domain_nossl)):
```
Comment (jpakkane) [inline mesonbuild/wrap/wrap.py]: And here.

**Comment by scivision** [comment]
This now allows URLS like mirror1.wrapdb.mesonbuild.com to pass, but not wrapdb.mesonbuild.com.evil.com

**Comment by scivision** [comment]
@jpakkane the CI actually passed, including cygwin gccx64ninja. Just a glitch in the GitHub azure connection.

File: mesonbuild/wrap/wrap.py [hunk 4]
```diff
@@ -53,35 +56,47 @@ def build_ssl_context() -> 'ssl.SSLContext':
...
+    if has_ssl and not url.scheme == 'https':
+        raise WrapException('WrapDB did not have expected SSL https url, instead got {}'.format(urlstr))
+    return url
+
 def open_wrapdburl(urlstring: str) -> 'http.client.HTTPResponse':
     global ssl_warning_printed
+
+    url = whitelist_wrapdb(urlstring)
     if has_ssl:
+        if not urlstring.startswith(whitelist_domain):
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[Flake8]__ undefined name 'whitelist_domain' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6238#issue-4898833))

<details>

| Rule |
| - |
| F821 |

</details>
<sub>posted by Sider</sub>

**Pull request merged**